File: 算法/代码随想录算法PDF/00调试源代码/337.打家劫舍-iii.cpp
/*
 * @lc app=leetcode.cn id=337 lang=cpp
 *
 * [337] 打家劫舍 III
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> robTree(TreeNode* root) {
        if (!root) return vector<int>{0, 0};
        vector<int> left = robTree(root->left);
        vector<int> right = robTree(root->right);
        int res1 = root->val + left[0] + right[0];
        int res2 = max(left[0], left[1]) + max(right[0], right[1]);
        return vector<int>{res2, res1};
    }
    int rob(TreeNode* root) {
        vector<int> res = robTree(root);
        return max(res[0], res[1]);
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/337.打家劫舍-iii.py
#
# @lc app=leetcode.cn id=337 lang=python3
#
# [337] 打家劫舍 III
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def robTree(self, root: Optional[TreeNode]) -> list:
        if not root: return [0, 0]
        left = self.robTree(root.left)
        right = self.robTree(root.right)
        res1 = root.val + left[0] + right[0]
        res2 = max(left) + max(right)
        return [res2, res1]
    def rob(self, root: Optional[TreeNode]) -> int:
        res = self.robTree(root)
        return max(res)
# @lc code=end

